File: dummy_thief/SocketWrapper.cpp
#include <iostream>

#include "SocketWrapper.h"

using namespace std;
using namespace boost;
using namespace boost::asio;

SocketWrapper::SocketWrapper(const string &server_ip, const string &rec_ip, 
	unsigned int port_send, unsigned int port_rec, 
	SocketReceiveCallback foo_callback) : foo_callback_(foo_callback), 
	port_send_(port_send), port_rec_(port_rec), server_ip_(server_ip), rec_ip_(rec_ip),
	_udp_sock(_serv, ip::udp::endpoint(ip::udp::v4(), port_rec_)),
	_resolver(_serv) {
		this->_start_recieve();
}

void SocketWrapper::run() {
	try {
		this->_isrunning = true;
		this->_serv.run();
	} catch (std::exception &e) {
		cerr << e.what() << endl;
	}
}

void SocketWrapper::send(const string& content) {
	_endp_rec_tosend = *_resolver.resolve({ip::udp::v4(), this->server_ip_, std::to_string(this->port_send_)});
	_SocketOperationCallback _foo_callback = [this, content] (const boost::system::error_code& error, std::size_t len) {
		if (!error) cout << "BAPol: data sended to " << this->server_ip_ << ":" << this->port_send_ << " : " << content << endl;
		this->_handle_send(error, len);
	};
	_udp_sock.async_send_to(buffer(content), _endp_rec_tosend, _foo_callback);
}

void SocketWrapper::_start_recieve() {
	_SocketOperationCallback _foo_callback = [this] (const boost::system::error_code& error, size_t len) {
		if (!error) cout << "BAPol: data received from " << this->_endp_rec_remote.address() << ":" << this->_endp_rec_remote.port() << " : " << string(begin(_rec_buffer), begin(_rec_buffer)+len) << endl;
		this->_handle_recieve(error, len);
	};
	_udp_sock.async_receive_from(buffer(this->_rec_buffer), _endp_rec_remote, _foo_callback);
}

void SocketWrapper::_handle_recieve(const boost::system::error_code& error, std::size_t len) {
	if (!error) {
		std::string _t_content(begin(_rec_buffer), begin(_rec_buffer)+len);
		std::string _t_endpoint;

		this->foo_callback_(_t_content, _t_endpoint);

		if (this->_isrunning) this->_start_recieve();
	}
}

File: dummy_thief/socket_wrapper.py
# python
# coding: UTF-8

import socket
import time

class SocketWrapper(object):

	_BUFSIZE = 4096

	def __init__(self, server_ip, port_send, rec_ip, port_rec, rec_callback):
		self.server_ip, self.server_port = server_ip, port_send
		self.self_ip, self.self_port = rec_ip, port_rec
		self._server_loc = (self.server_ip, self.server_port)
		self._rec_callback = rec_callback
		self._isrunning = False

		self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
		self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
		self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
		self._sock.bind((self.self_ip, self.self_port))

	def run(self):
		self._isrunning = True
		while self._isrunning:
			data, addr = self._sock.recvfrom(self._BUFSIZE)
			print("PyPol (%d): data received from %s : %s" % (time.time() * 1000, addr, data))
			self._rec_callback(data, addr)

	def send(self, content):
		self.sendto(content, self.server_ip, self.server_port)
		# print("PyPol (%d): data sended to %s : %s" % (time.time() * 1000, self._server_loc, content))
		# self._sock.sendto(content, self._server_loc)

	def sendto(self, content, addr, port):
		print("PyPol (%d): data sended to %s:%d : %s" % (time.time() * 1000, addr, port, content))
		self._sock.sendto(content, (addr, port))

	def pause(self):
		# self._sock.close()
		self._isrunning = False

	def close(self):
		self.pause()
		print("PyPol: closing socket...")
		self._sock.close()
